File: plugins/CircuitExplorer/io/BBPLoader.cpp
/* Copyright (c) 2015-2024, EPFL/Blue Brain Project
 * All rights reserved. Do not distribute without permission.
 * Responsible Author: Cyrille Favreau <cyrille.favreau@epfl.ch>
 *
 * This file is part of the circuit explorer for Brayns
 * <https://github.com/favreau/Brayns-UC-CircuitExplorer>
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License version 3.0 as published
 * by the Free Software Foundation.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include "BBPLoader.h"

#include <brayns/utils/Log.h>
#include <brayns/utils/Timer.h>
#include <brayns/utils/string/StringCase.h>

#include <api/ModelType.h>
#include <io/bbploader/CellLoader.h>
#include <io/bbploader/GIDLoader.h>
#include <io/bbploader/LoadContext.h>
#include <io/bbploader/ParameterCheck.h>
#include <io/bbploader/ReportLoader.h>
#include <io/bbploader/SynapseLoader.h>
#include <io/util/ProgressUpdater.h>

#include <filesystem>

namespace
{
class SynapseImporter
{
public:
    static void import(
        const bbploader::LoadContext &context,
        std::vector<std::shared_ptr<brayns::Model>> &modelList,
        ProgressUpdater &updater)
    {
        const auto &params = context.loadParameters;
        const auto afferent = params.load_afferent_synapses;
        const auto efferent = params.load_efferent_synapses;

        if (afferent)
        {
            updater.update();
            auto model = bbploader::SynapseLoader::load(context, true);
            modelList.push_back(std::move(model));
        }
        if (efferent)
        {
            updater.update();
            auto model = bbploader::SynapseLoader::load(context, false);
            modelList.push_back(std::move(model));
        }
    }
};
} // namespace

std::string BBPLoader::getName() const
{
    return "BBP loader";
}

std::vector<std::string> BBPLoader::getExtensions() const
{
    return {"BlueConfig", "CircuitConfig"};
}

std::vector<std::shared_ptr<brayns::Model>> BBPLoader::loadFile(const FileRequest &request)
{
    auto path = std::string(request.path);
    auto &callback = request.progress;
    auto &params = request.params;

    brayns::Timer timer;
    brayns::Log::info("[CE] {}: loading {}.", getName(), path);

    const brion::BlueConfig config(path);
    bbploader::ParameterCheck::checkInput(config, params);

    ProgressUpdater updater(callback, 3, 0.5f);

    const brain::Circuit circuit(config);
    const auto gids = bbploader::GIDLoader::compute(config, circuit, params);
    const bbploader::LoadContext context{circuit, gids, config, params};

    std::vector<std::shared_ptr<brayns::Model>> result;

    auto model = std::make_shared<brayns::Model>(ModelType::neurons);

    // Load neurons
    updater.beginStage("Neuron load", gids.size());
    auto compartments = bbploader::CellLoader::load(context, updater, *model);
    updater.endStage();

    // Load simulation
    updater.beginStage("Report load");
    bbploader::ReportLoader::load(context, compartments, updater, *model);
    updater.endStage();

    result.push_back(std::move(model));

    // Load synapses
    updater.beginStage("Synapse load", 2);
    SynapseImporter::import(context, result, updater);
    updater.endStage();

    updater.end("Generating rendering structures. Might take a while");

    brayns::Log::info("[CE] {}: Loaded {} model(s) in {} second(s).", getName(), result.size(), timer.seconds());

    return result;
}

File: python/brayns/plugins/bbp/bbp_loader.py
# Copyright (c) 2015-2022 EPFL/Blue Brain Project
# All rights reserved. Do not distribute without permission.
#
# Responsible Author: adrien.fleury@epfl.ch
#
# This file is part of Brayns <https://github.com/BlueBrain/Brayns>
#
# This library is free software; you can redistribute it and/or modify it under
# the terms of the GNU Lesser General Public License version 3.0 as published
# by the Free Software Foundation.
#
# This library is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this library; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, ClassVar

from brayns.core import Loader

from ..morphology import Morphology, serialize_morphology
from .bbp_cells import BbpCells
from .bbp_report import BbpReport


@dataclass
class BbpLoader(Loader):
    """BlueConfig / CircuitConfig loader.

    Use this loader to load circuit files with BBP internal format.

    :param cells: Cells to load, defaults to all.
    :type cells: BbpCells, optional
    :param report: Report to load, defaults to None.
    :type report: BbpReport | None, optional
    :param morphology: Specifies how to load the circuit morphologies.
    :type morphology: Morphology, optional
    :param load_afferent_synapses: Wether to load afferent, defaults to False.
    :type load_afferent_synapses: bool, optional
    :param load_efferent_synapses: Wether to load efferent, defaults to False.
    :type load_efferent_synapses: bool, optional
    """

    BLUE_CONFIG: ClassVar[str] = 'BlueConfig'
    CIRCUIT_CONFIG: ClassVar[str] = 'CircuitConfig'

    cells: BbpCells = field(default_factory=BbpCells.all)
    report: BbpReport | None = None
    morphology: Morphology = field(default_factory=Morphology)
    load_afferent_synapses: bool = False
    load_efferent_synapses: bool = False

    @classmethod
    @property
    def name(cls) -> str:
        return 'BBP loader'

    def get_properties(self) -> dict[str, Any]:
        return _serialize_loader(self)


def _serialize_loader(loader: BbpLoader) -> dict[str, Any]:
    message: dict[str, Any] = {
        'load_afferent_synapses': loader.load_afferent_synapses,
        'load_efferent_synapses': loader.load_efferent_synapses,
        **_serialize_cells(loader.cells),
        'neuron_morphology_parameters': serialize_morphology(loader.morphology),
    }
    if loader.report is not None:
        report = _serialize_report(loader.report)
        message.update(report)
    return message


def _serialize_cells(cells: BbpCells) -> dict[str, Any]:
    message = dict[str, Any]()
    if cells.density is not None:
        message['percentage'] = cells.density
    if cells.targets is not None:
        message['targets'] = cells.targets
    if cells.gids is not None:
        message['gids'] = cells.gids
    return message


def _serialize_report(report: BbpReport) -> dict[str, Any]:
    message: dict[str, Any] = {
        'report_type': report.type.value,
    }
    if report.name is not None:
        message['report_name'] = report.name
    if report.spike_transition_time is not None:
        message['spike_transition_time'] = report.spike_transition_time
    return message
